homeassistant/components/octoprint/sensor.py

```diff
@@ -22,118 +23,204 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        super().__init__(coordinator, sensor_name, "Current State")
+
+    @property
+    def state(self):
+        """Return sensor state."""
+        printer: OctoprintPrinterInfo = self.coordinator.data["printer"]
+        if not printer:
+            return None
+
+        return printer.state.text
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return "mdi:printer-3d"
+
+
+class OctoPrintJobPercentageSensor(OctoPrintSensorBase):
+    """Representation of an OctoPrint sensor."""
+
+    def __init__(self, coordinator: DataUpdateCoordinator, sensor_name: str):
+        """Initialize a new OctoPrint sensor."""
+        super().__init__(coordinator, sensor_name, "Job Percentage")
+
+    @property
+    def state(self):
+        """Return sensor state."""
+        job: OctoprintJobInfo = self.coordinator.data["job"]
+        if not job:
+            return 0
+
+        state = job.progress.completion
+        if not state:
+            return 0
+
+        return round(state, 2)
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return PERCENTAGE
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return "mdi:file-percent"
+
+
+class OctoPrintTimeRemainingSensor(OctoPrintSensorBase):
+    """Representation of an OctoPrint sensor."""
+
+    def __init__(self, coordinator: DataUpdateCoordinator, sensor_name: str):
+        """Initialize a new OctoPrint sensor."""
+        super().__init__(coordinator, sensor_name, "Time Remaining")
+
+    @property
+    def state(self):
+        """Return sensor state."""
+        job: OctoprintJobInfo = self.coordinator.data["job"]
+        if not job:
+            return None
+
+        return job.progress.print_time_left
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return TIME_SECONDS
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return "mdi:clock-start"
+
+
+class OctoPrintTimeElapsedSensor(OctoPrintSensorBase):
+    """Representation of an OctoPrint sensor."""
+
+    def __init__(self, coordinator: DataUpdateCoordinator, sensor_name: str):
+        """Initialize a new OctoPrint sensor."""
+        super().__init__(coordinator, sensor_name, "Time Elapsed")
+
     @property
     def state(self):
-        """Return the state of the sensor."""
-        sensor_unit = self.unit_of_measurement
-        if sensor_unit in (TEMP_CELSIUS, PERCENTAGE):
-            # API sometimes returns null and not 0
-            if self._state is None:
-                self._state = 0
-            return round(self._state, 2)
-        return self._state
+        """Return sensor state."""
+        job: OctoprintJobInfo = self.coordinator.data["job"]
+        if not job:
+            return None
+
+        return job.progress.print_time
```

Comment:
This should be a timestamp device class sensor. Calculate an absolute UTC timestamp and return an isoformatted datetime string.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

Reply:
Given this would be a breaking change, would I be OK to change the sensor names as well? Time Elapsed => Start Time and Time Remaining => Estimated End Time?

Reply:
Sounds good!